File: crates/token-saver-core/src/skeleton.rs
//! AST Skeletonizer & Symbol Extractor in Rust.
//!
//! Replaces function and method bodies with '...' to extract structural APIs,
//! saving 70-90% of tokens when exploring source files.

use std::path::Path;
use crate::parser::{detect_language, parse_code, SupportedLanguage};
use crate::token_counter::estimate_tokens;

#[derive(Debug, Clone)]
struct ReplaceRange {
    start_byte: usize,
    end_byte: usize,
    replacement: &'static str,
}

/// Builds a skeleton of source code by replacing function/method bodies with '...'.
pub fn build_skeleton(source_code: &str, lang: SupportedLanguage) -> String {
    let tree = match parse_code(source_code, lang) {
        Some(t) => t,
        None => return source_code.to_string(),
    };

    let root_node = tree.root_node();
    let mut replace_ranges: Vec<ReplaceRange> = Vec::new();

    fn walk(node: tree_sitter::Node, lang: SupportedLanguage, ranges: &mut Vec<ReplaceRange>) {
        let node_kind = node.kind();

        match lang {
            SupportedLanguage::Python => {
                if node_kind == "function_definition" {
                    let mut body = None;
                    for child in node.children(&mut node.walk()) {
                        if child.kind() == "block" {
                            body = Some(child);
                            break;
                        }
                    }

                    if let Some(b) = body {
                        let mut start_byte = b.start_byte();
                        // Preserve Python docstring if present as first statement
                        if b.child_count() > 0 {
                            let first_stmt = b.child(0);
                            if let Some(stmt) = first_stmt {
                                if stmt.kind() == "expression_statement" && stmt.child_count() > 0 {
                                    if let Some(expr) = stmt.child(0) {
                                        if expr.kind() == "string" {
                                            start_byte = stmt.end_byte();
                                        }
                                    }
                                }
                            }
                        }

                        if start_byte < b.end_byte() {
                            ranges.push(ReplaceRange {
                                start_byte,
                                end_byte: b.end_byte(),
                                replacement: " ...\n",
                            });
                        }
                    }
                }
            }
            SupportedLanguage::Rust => {
                if node_kind == "function_item" {
                    let mut body = None;
                    for child in node.children(&mut node.walk()) {
                        if child.kind() == "block" {
                            body = Some(child);
                            break;
                        }
                    }
                    if let Some(b) = body {
                        if b.start_byte() + 1 < b.end_byte() {
                            ranges.push(ReplaceRange {
                                start_byte: b.start_byte() + 1,
                                end_byte: b.end_byte() - 1,
                                replacement: "\n    ...\n",
                            });
                        }
                    }
                }
            }
            SupportedLanguage::JavaScript
            | SupportedLanguage::TypeScript
            | SupportedLanguage::C
            | SupportedLanguage::Cpp
            | SupportedLanguage::Java
            | SupportedLanguage::CSharp
            | SupportedLanguage::Php
            | SupportedLanguage::Bash => {
                if matches!(
                    node_kind,
                    "function_declaration"
                        | "method_definition"
                        | "arrow_function"
                        | "method_declaration"
                        | "function_definition"
                ) {
                    let mut body = None;
                    for child in node.children(&mut node.walk()) {
                        if matches!(child.kind(), "statement_block" | "compound_statement" | "block") {
                            body = Some(child);
                            break;
                        }
                    }
                    if let Some(b) = body {
                        if b.start_byte() + 1 < b.end_byte() {
                            ranges.push(ReplaceRange {
                                start_byte: b.start_byte() + 1,
                                end_byte: b.end_byte() - 1,
                                replacement: "\n  ...\n",
                            });
                        }
                    }
                }
            }
            SupportedLanguage::Ruby => {
                if matches!(node_kind, "method" | "singleton_method") {
                    let mut body = None;
                    for child in node.children(&mut node.walk()) {
                        if child.kind() == "body_statement" {
                            body = Some(child);
                            break;
                        }
                    }
                    if let Some(b) = body {
                        if b.start_byte() < b.end_byte() {
                            ranges.push(ReplaceRange {
                                start_byte: b.start_byte(),
                                end_byte: b.end_byte(),
                                replacement: "\n    ...\n  ",
                            });
                        }
                    }
                }
            }
            SupportedLanguage::Html | SupportedLanguage::Css | SupportedLanguage::Json => {}
            SupportedLanguage::Go => {
                if matches!(node_kind, "function_declaration" | "method_declaration") {
                    let mut body = None;
                    for child in node.children(&mut node.walk()) {
                        if child.kind() == "block" {
                            body = Some(child);
                            break;
                        }
                    }
                    if let Some(b) = body {
                        if b.start_byte() + 1 < b.end_byte() {
                            ranges.push(ReplaceRange {
                                start_byte: b.start_byte() + 1,
                                end_byte: b.end_byte() - 1,
                                replacement: "\n\t...\n",
                            });
                        }
                    }
                }
            }
        }

        // Recursively traverse children
        for child in node.children(&mut node.walk()) {
            walk(child, lang, ranges);
        }
    }

    walk(root_node, lang, &mut replace_ranges);

    if replace_ranges.is_empty() {
        return source_code.to_string();
    }

    // Sort by start_byte
    replace_ranges.sort_by_key(|r| r.start_byte);

    // Resolve overlapping ranges (keep outermost)
    let mut filtered_ranges: Vec<ReplaceRange> = Vec::new();
    for r in replace_ranges {
        if let Some(last) = filtered_ranges.last() {
            if r.start_byte < last.end_byte {
                continue;
            }
        }
        filtered_ranges.push(r);
    }

    // Splice string bytes
    let source_bytes = source_code.as_bytes();
    let mut result_bytes = Vec::new();
    let mut last_end = 0;

    for r in filtered_ranges {
        if r.start_byte >= last_end && r.end_byte <= source_bytes.len() {
            result_bytes.extend_from_slice(&source_bytes[last_end..r.start_byte]);
            result_bytes.extend_from_slice(r.replacement.as_bytes());
            last_end = r.end_byte;
        }
    }

    if last_end < source_bytes.len() {
        result_bytes.extend_from_slice(&source_bytes[last_end..]);
    }

    String::from_utf8(result_bytes).unwrap_or_else(|_| source_code.to_string())
}

/// Finds the full source implementation of a symbol (function, method, class, struct).
pub fn find_symbol_in_code(source_code: &str, lang: SupportedLanguage, symbol_name: &str) -> Option<String> {
    let tree = parse_code(source_code, lang)?;
    let source_bytes = source_code.as_bytes();

    fn walk(
        node: tree_sitter::Node,
        source_bytes: &[u8],
        target_name: &str,
    ) -> Option<String> {
        let node_kind = node.kind();
        let is_target_def = matches!(
            node_kind,
            "function_definition"
                | "class_definition"
                | "function_declaration"
                | "class_declaration"
                | "method_definition"
                | "function_item"
                | "struct_item"
                | "enum_item"
                | "impl_item"
        );

        if is_target_def {
            for child in node.children(&mut node.walk()) {
                if matches!(
                    child.kind(),
                    "identifier" | "type_identifier" | "property_identifier" | "name"
                ) {
                    if let Ok(name_str) = std::str::from_utf8(&source_bytes[child.start_byte()..child.end_byte()]) {
                        if name_str == target_name {
                            if let Ok(full_impl) = std::str::from_utf8(&source_bytes[node.start_byte()..node.end_byte()]) {
                                return Some(full_impl.to_string());
                            }
                        }
                    }
                }
            }
        }

        for child in node.children(&mut node.walk()) {
            if let Some(found) = walk(child, source_bytes, target_name) {
                return Some(found);
            }
        }

        None
    }

    walk(tree.root_node(), source_bytes, symbol_name)
}

/// High-level function: extracts code skeleton from a file on disk.
pub fn get_code_skeleton_file<P: AsRef<Path>>(file_path: P) -> String {
    let p = file_path.as_ref();
    if !p.exists() {
        return format!("Error: File {} not found.", p.display());
    }
    if p.is_dir() {
        return format!("Error: '{}' is a directory, not a file.", p.display());
    }

    let content = match std::fs::read(p) {
        Ok(bytes) => match String::from_utf8(bytes) {
            Ok(s) => s,
            Err(e) => String::from_utf8_lossy(e.as_bytes()).into_owned(),
        },
        Err(e) => return format!("Error reading file {}: {}", p.display(), e),
    };

    let lang = match detect_language(p) {
        Some(l) => l,
        None => return content, // Return full text if language not supported
    };

    let skeleton = build_skeleton(&content, lang);
    let orig_tokens = estimate_tokens(&content);
    let skel_tokens = estimate_tokens(&skeleton);
    let savings_pct = if orig_tokens > 0 {
        ((orig_tokens.saturating_sub(skel_tokens) as f64 / orig_tokens as f64) * 100.0) as usize
    } else {
        0
    };

    let comment_prefix = if lang == SupportedLanguage::Python { "#" } else { "//" };
    format!("{skeleton}\n{comment_prefix} Token-Saver: {orig_tokens} -> {skel_tokens} tokens ({savings_pct}% saved)")
}

/// High-level function: extracts a specific symbol from a file on disk.
pub fn get_symbol_file<P: AsRef<Path>>(file_path: P, symbol_name: &str) -> String {
    let p = file_path.as_ref();
    if !p.exists() {
        return format!("Error: File {} not found.", p.display());
    }
    if p.is_dir() {
        return format!("Error: '{}' is a directory, not a file.", p.display());
    }

    let content = match std::fs::read(p) {
        Ok(bytes) => match String::from_utf8(bytes) {
            Ok(s) => s,
            Err(e) => String::from_utf8_lossy(e.as_bytes()).into_owned(),
        },
        Err(e) => return format!("Error reading file {}: {}", p.display(), e),
    };

    let lang = match detect_language(p) {
        Some(l) => l,
        None => return format!("Error: Could not detect language for {}.", p.display()),
    };

    match find_symbol_in_code(&content, lang, symbol_name) {
        Some(impl_str) => impl_str,
        None => format!("Symbol '{symbol_name}' not found in {}.", p.display()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_python_skeleton() {
        let code = r#"def calculate_total(price: float, tax: float) -> float:
    """Calculates total with tax."""
    val = price * (1.0 + tax)
    return val

def helper():
    return 42
"#;
        let skel = build_skeleton(code, SupportedLanguage::Python);
        assert!(skel.contains("def calculate_total(price: float, tax: float) -> float:"));
        assert!(skel.contains(r#""""Calculates total with tax.""""#));
        assert!(skel.contains("..."));
        assert!(!skel.contains("val = price * (1.0 + tax)"));
        assert!(!skel.contains("return 42"));
    }

    #[test]
    fn test_rust_skeleton() {
        let code = r#"pub fn compute(a: i32, b: i32) -> i32 {
    let res = a + b * 2;
    res
}
"#;
        let skel = build_skeleton(code, SupportedLanguage::Rust);
        assert!(skel.contains("pub fn compute(a: i32, b: i32) -> i32 {"));
        assert!(skel.contains("..."));
        assert!(!skel.contains("let res = a + b * 2;"));
    }

    #[test]
    fn test_find_symbol_python() {
        let code = r#"def alpha():
    return 1

def beta(x: int) -> int:
    return x * 2
"#;
        let found = find_symbol_in_code(code, SupportedLanguage::Python, "beta").expect("Should find beta");
        assert!(found.starts_with("def beta(x: int) -> int:"));
        assert!(found.contains("return x * 2"));
        assert!(!found.contains("alpha"));
    }
}

File: src/token_saver/tools/skeleton.py
from __future__ import annotations

import os
from typing import Any

from token_saver.parsers.languages import parse_code
from token_saver.utils.file_utils import detect_language, read_file_text
from token_saver.utils.token_counter import estimate_tokens


def _build_skeleton(source_code: str, language: str) -> str:
    """Builds a skeleton of the source code by replacing function/method bodies with '...'."""
    tree = parse_code(source_code, language)
    if not tree:
        return source_code

    root_node = tree.root_node

    # We will collect byte ranges to replace with '...'
    replace_ranges = []

    def walk(node):
        node_type = node.type

        # Identify function-like bodies to replace
        if language == "python":
            if node_type == "function_definition":
                body = None
                for child in node.children:
                    if child.type == "block":
                        body = child
                        break
                if body:
                    # Check for docstring
                    start_byte = body.start_byte
                    first_stmt = body.children[0] if body.children else None
                    if first_stmt and first_stmt.type == "expression_statement":
                        if first_stmt.children and first_stmt.children[0].type == "string":
                            start_byte = first_stmt.end_byte

                    if start_byte < body.end_byte:
                        replace_ranges.append((start_byte, body.end_byte, " ...\n"))

        elif language == "ruby":
            if node_type in ["method", "singleton_method"]:
                body = None
                for child in node.children:
                    if child.type == "body_statement":
                        body = child
                        break
                if body and body.start_byte < body.end_byte:
                    replace_ranges.append((body.start_byte, body.end_byte, "\n    ...\n  "))

        elif language in [
            "javascript",
            "typescript",
            "tsx",
            "jsx",
            "go",
            "java",
            "c",
            "cpp",
            "rust",
            "c_sharp",
            "php",
            "kotlin",
        ]:
            # Generic body replacement for C-family and block-based languages
            if node_type in [
                "function_declaration",
                "method_definition",
                "arrow_function",
                "method_declaration",
                "function_definition",
                "function_item",
            ]:
                body = None
                for child in node.children:
                    if child.type in ["statement_block", "block", "compound_statement", "function_body"]:
                        body = child
                        break

                if body and body.start_byte < body.end_byte - 1:
                    # Keep braces if possible
                    replace_ranges.append((body.start_byte + 1, body.end_byte - 1, "\n  ...\n"))

        # Traverse children
        for child in node.children:
            walk(child)

    walk(root_node)

    if not replace_ranges:
        return source_code

    # Sort ranges by start_byte
    replace_ranges.sort(key=lambda x: x[0])

    # Resolve overlapping ranges (keep the outermost)
    filtered_ranges = []
    for r in replace_ranges:
        if not filtered_ranges:
            filtered_ranges.append(r)
        else:
            last_r = filtered_ranges[-1]
            if r[0] < last_r[1]:
                continue
            else:
                filtered_ranges.append(r)

    # Reconstruct the string
    source_bytes = source_code.encode("utf-8")
    result_bytes = bytearray()
    last_end = 0

    for start, end, replacement in filtered_ranges:
        result_bytes.extend(source_bytes[last_end:start])
        result_bytes.extend(replacement.encode("utf-8"))
        last_end = end

    result_bytes.extend(source_bytes[last_end:])

    return result_bytes.decode("utf-8")


def _find_symbol(source_code: str, language: str, symbol_name: str) -> str:
    """Finds and returns the full text of a symbol by name."""
    tree = parse_code(source_code, language)
    if not tree:
        return ""

    root_node = tree.root_node

    def walk(node):
        # Look for identifiers that match the symbol name
        if node.type in [
            "function_definition",
            "class_definition",
            "function_declaration",
            "class_declaration",
            "method_definition",
            "method_declaration",
            "function_item",
            "struct_item",
            "method",
            "singleton_method",
            "class",
            "module",
        ]:
            name_node = None
            for child in node.children:
                if child.type in [
                    "identifier",
                    "type_identifier",
                    "property_identifier",
                    "name",
                    "simple_identifier",
                    "constant",
                ]:
                    name_node = child
                    break

            if (
                name_node
                and source_code.encode("utf-8")[name_node.start_byte : name_node.end_byte].decode("utf-8")
                == symbol_name
            ):
                return source_code.encode("utf-8")[node.start_byte : node.end_byte].decode("utf-8")

        for child in node.children:
            res = walk(child)
            if res:
                return res
        return None

    res = walk(root_node)
    return res if res else f"Symbol '{symbol_name}' not found."


def get_code_skeleton(file_path: str) -> str:
    """
    Extracts a structural skeleton from a source file.
    Use this INSTEAD of reading full files when you need to understand a file's API/structure.
    It keeps imports, classes, function signatures, and docstrings, but replaces bodies with '...'.
    """
    if not os.path.exists(file_path):
        return f"Error: File {file_path} not found."

    content = read_file_text(file_path)
    language = detect_language(file_path)

    if not language:
        return content

    skeleton = _build_skeleton(content, language)

    orig_tokens = estimate_tokens(content)
    skel_tokens = estimate_tokens(skeleton)
    savings_pct = 0
    if orig_tokens > 0:
        savings_pct = int(((orig_tokens - skel_tokens) / orig_tokens) * 100)

    skeleton += f"\n# Token-Saver: {orig_tokens} → {skel_tokens} tokens ({savings_pct}% saved)"
    try:
        from token_saver.telemetry.stats import tracker

        tracker.record_savings("skeleton", orig_tokens, skel_tokens)
    except Exception:
        pass
    return skeleton


def get_symbol(file_path: str, symbol_name: str) -> str:
    """
    Extracts the FULL implementation of a specific function, method, or class from a file by name.
    Useful after viewing a skeleton to fetch only the specific function needed.
    """
    if not os.path.exists(file_path):
        return f"Error: File {file_path} not found."

    content = read_file_text(file_path)
    language = detect_language(file_path)

    if not language:
        return f"Error: Could not detect language for {file_path}."

    return _find_symbol(content, language, symbol_name)


def register_skeleton_tools(mcp: Any) -> None:
    """Registers skeleton tools with the FastMCP application."""

    @mcp.tool(
        description="Extracts a structural skeleton from a source file, replacing bodies with '...'. Use this INSTEAD of reading full files to understand structure."
    )
    def tool_get_code_skeleton(file_path: str) -> str:
        return get_code_skeleton(file_path)

    @mcp.tool(
        description="Extracts the FULL implementation of a specific function, method, or class from a file by name."
    )
    def tool_get_symbol(file_path: str, symbol_name: str) -> str:
        return get_symbol(file_path, symbol_name)
